provisioning/baremetal_webhook.go

```diff
@@ -0,0 +1,159 @@
+	if !info.BaremetalWebhookEnabled {
+		// In case service-ca operator is degraded and we detect it in reconciliation loop
+		// We should delete validatingwebhook resources(ValidatingWebhookConfiguration, Service)
+		// to allow BMO continues serving. If we detect in first run, it is safe to run these
+		// deletions.
+		err := DeleteValidatingWebhook(info)
+		return false, err
+	}
+
+	webhookService := newBaremetalOperatorWebhookService(info.Namespace)
+	_, _, err := resourceapply.ApplyService(info.Client.CoreV1(), info.EventRecorder, webhookService)
+	if err != nil {
+		err = errors.Wrap(err, "unable to create validatingwebhook service")
+		return false, err
+	}
+
+	vw := newBaremetalOperatorWebhook(info.Namespace)
+	expectedGeneration := resourcemerge.ExpectedValidatingWebhooksConfiguration(validatingWebhookConfigurationName, info.ProvConfig.Status.Generations)
+	validatingWebhook, updated, err := resourceapply.ApplyValidatingWebhookConfiguration(info.Client.AdmissionregistrationV1(), info.EventRecorder, vw, expectedGeneration)
+	if err != nil {
+		err = errors.Wrap(err, "unable to create validatingwebhook configuration")
+		return false, err
+	}
+
+	if updated {
+		resourcemerge.SetValidatingWebhooksConfigurationGeneration(&info.ProvConfig.Status.Generations, validatingWebhook)
+	}
+
+	return updated, nil
+}
+
+// BaremetalWebhookDependenciesReady checks dependencies to enable Baremetal
+// Operator ValidatingWebhook.
+func BaremetalWebhookDependenciesReady(client osclientset.Interface) bool {
+	// Service CA operator will inject required certificates.
+	// If Service CA is not ready, ValidatingWebhook should not be enabled.
+	return serviceCAOperatorReady(client)
+}
+
+// DeleteValidatingWebhook deletes ValidatingWebhookConfiguration and
+// service resources.
+func DeleteValidatingWebhook(info *ProvisioningInfo) error {
+	err := client.IgnoreNotFound(info.Client.CoreV1().Services(info.Namespace).Delete(context.Background(), validatingWebhookService, metav1.DeleteOptions{}))
+	if err != nil {
+		return err
+	}
+
+	err = client.IgnoreNotFound(info.Client.AdmissionregistrationV1().
+		ValidatingWebhookConfigurations().
+		Delete(context.Background(), validatingWebhookConfigurationName, metav1.DeleteOptions{}))
+
+	return err
+}
+
+func newBaremetalOperatorWebhookService(targetNamespace string) *corev1.Service {
+	webhookPort, _ := strconv.ParseInt(baremetalWebhookPort, 10, 32) // #nosec
+	return &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      validatingWebhookService,
+			Namespace: targetNamespace,
+			Annotations: map[string]string{
+				"include.release.openshift.io/self-managed-high-availability": "true",
+				"include.release.openshift.io/single-node-developer":          "true",
+				"service.beta.openshift.io/serving-cert-secret-name":          baremetalWebhookSecretName,
+			},
+			Labels: map[string]string{
+				"k8s-app":    metal3AppName,
+				cboLabelName: stateService,
```

Comment:
I don't want to use this label as the target. The purpose of the metal3-state service is to point to the Pod that holds the DB and httpd. It exists specifically so that things that are stateless, like baremetal-operator, can eventually be _removed_ from that Pod.
If we want a service to point at the baremetal-operator, we should create another label for it so that they can move independently in future.

Reply:
Thank you @zaneb, that is a good point and I'll create another label.